src/Streamly/Coreutils/Uniq.hs

```diff
@@ -0,0 +1,217 @@
+--
+-- @since 0.1.0.0
+compareIgnCase :: String -> String -> Bool
+compareIgnCase str1 str2 | length str1 /= length str2 = False
+compareIgnCase (x1:xs) (y1:ys) =
+    (toLower x1 == toLower y1) && compareIgnCase xs ys
+compareIgnCase [] [] = True
+
+
+
+-- |
+-- Produces a comparison function of type @String -> String -> Bool@
+-- using the options specified
+--
+-- @since 0.1.0.0
+compareUsingOptions :: UniqOptions -> (String -> String -> Bool)
+compareUsingOptions opt = compareXYZ (ignoreCase opt) (skipFields opt) (skipChar opt) (checkChar opt)
+   where
+      -- ^
+      -- @compareXYZ ign x y z s1 s2@ does the following in order:
+      -- skips first @x@ non   -space fields,
+      -- skips first @y@ characters and takes first @z@ characters
+      -- from the strings @s1@ and @s2@ and returns True only if they are equal
+      -- after the above operations. Ignores case if
+      -- @ign@ case is True
+      --
+      -- @since 0.1.0.0
+      compareXYZ :: Bool -> Int -> Int -> Maybe Int -> String -> String -> Bool
+      compareXYZ ignCase x y mz str1 str2 =
+         let
+            ix1 = indexAfterSkippingNWords x str1
+            ix2 = indexAfterSkippingNWords x str2
+            compareFunc =
+               if ignCase
+               then compareIgnCase
+               else (==)
+            compareWith :: (String -> String -> Bool) -> String -> String -> Bool
+            compareWith eq s1 s2 = eq s1 s2
+            -- ^
+            -- Gives the number of characters to be deleted
+            -- from the beginning of the string to skip @n@
+            -- non-space fields\/strings
+            --
+            -- @since 0.1.0.0
+            indexAfterSkippingNWords :: Int -> String -> Int
+            indexAfterSkippingNWords n str = go 0 n
+               where
+                  len = length str
+                  go i 0 = i
+                  go i _ | i >= (len - 1) = len
+                  go i xx = if isChar (str !! i) && isSpace (str !! (i + 1))
+                           then go (i + 1) (xx - 1)
+                           else go (i + 1) xx
+                  isChar = not . isSpace
+         in
+            case mz of
+               Just z -> compareWith
+                         compareFunc
+                         (slice (ix1 + y) z str1)
+                         (slice (ix2 + y) z str2)
+               _  -> compareWith
+                     compareFunc
+                     (drop (ix1 + y) str1)
+                     (drop (ix2 + y) str2)
+
+
+-- |
+-- Applies the comparison function to adjacent strings in the stream
+-- and returns a @Count Int String@ stream where the integer
+-- is the count of occurrences of the string
+--
+-- @since 0.1.0.0
+getRepetition :: (IsStream t, Monad m) => (String -> String -> Bool) -> t m String -> t m UniqResult
+getRepetition comparator = S.groupsBy comparator
+                              (FL.mkPureId (\(Count i a) s ->
+                                  if a == ""
```

Comment:
Why do you need this if statement? Just `Count (i + 1) s` would be sufficient?

Reply:
Actually, after comparison, if two strings are found the same, the previous and original one is taken. That's why.

Reply:
> the previous and original one is taken

What?

Reply:
Let's say with `skip fields = 1, two strings "field1 ye" and "field2 ye"` would be declared the same by the comparator. Then the stream should have `UniqResult 2 "field1 ye"` (first and original)